src/circuit_proof/mod.rs

```diff
@@ -0,0 +1,713 @@
+#![allow(non_snake_case)]
+
+use rand::{CryptoRng, Rng};
+use std::iter;
+
+use curve25519_dalek::ristretto::RistrettoPoint;
+use curve25519_dalek::scalar::Scalar;
+use curve25519_dalek::traits::{IsIdentity, MultiscalarMul, VartimeMultiscalarMul};
+
+use generators::Generators;
+use proof_transcript::ProofTranscript;  
+use util;
+use inner_product_proof::{inner_product, InnerProductProof};
+
+#[derive(Clone, Debug)]
+pub struct CircuitProof {
+    pub A_I: RistrettoPoint,
+    pub A_O: RistrettoPoint,
+    pub S: RistrettoPoint,
+    pub T_1: RistrettoPoint,
+    pub T_3: RistrettoPoint,
+    pub T_4: RistrettoPoint,
+    pub T_5: RistrettoPoint,
+    pub T_6: RistrettoPoint,
+    pub t_x: Scalar,
+    pub t_x_blinding: Scalar,
+    pub e_blinding: Scalar,
+    pub l_vec: Vec<Scalar>,
+    pub r_vec: Vec<Scalar>,
+    pub ipp_proof: InnerProductProof,
+}
+
+impl CircuitProof {
+    // XXX removed `c` (constants vector) from prover input, since prover doesn't use it to create proof.
+    pub fn prove<R: Rng + CryptoRng>(
+        gen: &Generators,
+        transcript: &mut ProofTranscript,
+        rng: &mut R,
+    
+        n: usize,
+        m: usize,
+        q: usize,
+
+        W_L: Vec<Vec<Scalar>>, // Q vectors, of length n each
+        W_R: Vec<Vec<Scalar>>,
+        W_O: Vec<Vec<Scalar>>,
+        W_V: Vec<Vec<Scalar>>, // Q vectors, of length m each
+        a_L: Vec<Scalar>,
+        a_R: Vec<Scalar>,
+        a_O: Vec<Scalar>,
+        v_blinding: Vec<Scalar>,
```

Comment:
All of `n, m, q, W_L, W_R, W_O, W_V` are circuit parameters, while the `a_L, a_R, a_O` are derived from the witness, right?

Does it make sense to split out the non-witness-related parameters into a single `Circuit` object, instead of having the callers pass all of them manually?

Reply:
that is correct - yes, that makes it a lot cleaner! Will do